### PR TITLE
Stats: Enable date ranges on Summary pages

### PR DESCRIPTION
**Pull request by a8ck3n**
Related to: https://github.com/Automattic/red-team/issues/252

## Proposed Changes

Updates the Traffic page to push the full date range to the Summary pages.

Updates the Summary pages to accept the `endDate` parameter. If provided, validates the date and creates a dateRange object that is passed into the summary page. This is then used to update the segmented control (the row of connected buttons), the summary header, and in the API request.

## Why are these changes being made?

Part of the date filtering project.

pejTkB-1zs-p2

## Testing Instructions

* Open the Calypso Live link.
* Test a few of the "View all" links on the Traffic page to confirm the resulting summary pages still work as expected. They should show a summary based on the current period from the Traffic page: day, week, month, etc.
* Enable the new date filtering: `&flags=stats/new-date-filtering`
* Hover over a "View all" link and confirm that both `startDate` and `endDate` values are present in the link.
* Click through to the summary page and confirm the custom range us applied to the page. (see screenshot)
* Click around to see if anything seems broken!

### New endDate param added to summary links
SCR-20241123-bzlm

### Updated summary page respecting custom range
SCR-20241123-cafz

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/stats/stats-date-picker/index.jsx [hunk 1]
```diff
@@ -51,6 +56,21 @@ class StatsDatePicker extends Component {
 		}
 	}
 
+	dateForCustomRange() {
+		const { query, moment, translate } = this.props;
+
+		const localizedStartDate = moment( query.start_date );
+		const localizedEndDate = moment( query.date );
+
+		return translate( '%(startDate)s ~ %(endDate)s', {
```
Comment (kangzj) [inline client/my-sites/stats/stats-date-picker/index.jsx:65]: We should be able to leverage the changes from @annacmc which just got merged

https://github.com/Automattic/wp-calypso/pull/96661/files#diff-7b415fb32f446cf84f379854d3e2a15a55564aae14c3f9eb1f5cf404bdb910bfL55

File: client/my-sites/stats/stats-module/all-time-nav.jsx [hunk 2]
```diff
@@ -62,12 +65,21 @@ export const StatsModuleSummaryLinks = ( props ) => {
 		recordStats( item );
 	};
 
+	// Template for standard range options (7, 30, Quarter, Year, All Time).
 	const summaryPath = `/stats/day/${ path }/${ siteSlug }?startDate=${ moment().format(
```
Comment (kangzj) [inline client/my-sites/stats/stats-module/all-time-nav.jsx:69]: Not related to the PR, but again we should always use `momentSiteZone` for today until any further decision is made. cc @Automattic/red

File: client/my-sites/stats/stats-module/all-time-nav.jsx [hunk 3]
```diff
@@ -62,12 +65,21 @@ export const StatsModuleSummaryLinks = ( props ) => {
...
-	const summaryPeriodPath = `/stats/${
+
+	// Path for summary or custom range option. ie: The first button in the row.
+	// Defaults to one day/week/month/year.
+	let summaryPeriodPath = `/stats/${
 		period.period
 	}/${ path }/${ siteSlug }?startDate=${ period.endOf.format( 'YYYY-MM-DD' ) }`;
+	// Override if custom range was used in query.
+	if ( query.start_date ) {
+		summaryPeriodPath = `/stats/${ period.period }/${ path }/${ siteSlug }?startDate=${ query.start_date }&endDate=${ query.date }`;
```
Comment (kangzj) [inline client/my-sites/stats/stats-module/all-time-nav.jsx:80]: nitpick - the startDate here means a different date than the original param, so I wonder if we should use a different name. Maybe to be consistent with the Traffic page and use `chartStart` and `chartEnd`?

Same applies to `getSummaryLink` and `getStatHref`.

Plus it becomes fairly confusing when query is using `start_date` + `date` combination.

**Review by kangzj — APPROVED** [review]
It works as described. Good job!

One nitpick is that when clicking on another tab for example `7 day`, you wouldn't be able to go back to the Custom Date Range anymore. I'm not sure what would be the best behaviour here and it wouldn't need urgent care.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17017397

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @a8ck3n for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ck3n** [comment]
> One nitpick is that when clicking on another tab for example `7 day`, you wouldn't be able to go back to the Custom Date Range anymore. I'm not sure what would be the best behaviour here and it wouldn't need urgent care.

Yeah, I wasn't too sure about this one. The browser's back button will work but otherwise, it feels a bit awkward to keep the custom date range in the control without presenting the actual dates anywhere. My thought was to get it working and then see what the feedback is from design. Eventually we'd want to just remove this control and replace it with the date picker though, right?

**Comment by kangzj** [comment]
> My thought was to get it working and then see what the feedback is from design. Eventually we'd want to just remove this control and replace it with the date picker though, right?

I agree!

**Pull request merged**